crates/prune/src/pruner.rs

```diff
@@ -0,0 +1,168 @@
+//! Support for pruning.
+
+use crate::PrunerError;
+use futures_util::Stream;
+use reth_primitives::BlockNumber;
+use reth_provider::CanonStateNotification;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tracing::debug;
+
+/// The future that returns the owned pipeline and the result of the pipeline run. See
+/// [Pruner::run_as_fut].
+pub type PrunerFut<St> = Pin<Box<dyn Future<Output = PrunerWithResult<St>> + Send>>;
```

Comment:
not a fan of this pattern, we had to do it for the pipeline because of its blocking behavior which makes it not pollable.

But I don't really understand why we need to spawn and stop the pruner task?

hmm, if we spawn this, how long does it run then?

Reply:
> But I don't really understand why we need to spawn and stop the pruner task?

1. It needs to acquire the write tx that shouldn't be shared by any other tasks, so we need one place to control who owns the write tx.
2. It's a blocking task, same as pipeline because it interacts with database.